### PR TITLE
[flang] Improve alias analysis to be precise for box and box.base_addr

### PR DESCRIPTION
**Pull request by razvanlupusoru**
After PR#68727 the source for both the fir.box_addr and a box became the same. Thus the detection that only one of the sources was direct and the special logic around it was being skipped. As a result, the test included would show a "MayAlias" result instead of a "NoAlias" result.

### CODE REVIEW

**Review by tblah — APPROVED** [review]
This makes sense to me, but please wait for approval from at least one of the other reviewers.

Out of interest, are you able to share what you are using this for? I believe the existing alias analysis infrastructure should already be able to distinguish between box and non-box accesses.

**Comment by razvanlupusoru** [comment]
> Out of interest, are you able to share what you are using this for?
I was trying it out on some OpenACC. I am interested in creating an alias analysis framework (at MLIR level) for it as well.

> I believe the existing alias analysis infrastructure should already be able to distinguish between box and non-box accesses.
Are you referring to the TBAA implementation? As far as I understood this was mostly for passing information to LLVM, not for alias analysis in MLIR passes.

**Review by Renaud-K — APPROVED** [review]

**Comment by tblah** [comment]
> Are you referring to the TBAA implementation? As far as I understood this was mostly for passing information to LLVM, not for alias analysis in MLIR passes.

Yes that was what I meant. Yes it is not for MLIR passes.

**Pull request merged**
